### ISSUE TITLE
Pairdrop

### ISSUE THREAD

**Issue by ueen** [fm-sys/snapdrop-android#269]
### Describe the feature

Hey, just fixing my Addon and Snapdrop never works (offline), i was just wondering, would you consider supporting pairdrop?
Its incompatible as of right know and i`m unsure how much work it would be (think not to much but still some), or if it would make sense to fork and do a seperat Pairdrop App/Addon, just wanted to coordinate so i can use the app and my addon in unity :D

Pairdrop seems promising with new features and a semingly engaged maintainer and uptime in the last week...

Interested in what you think :)

Maybe @schlagmichdoch can weigh in if they would want that 

### Describe alternatives you've considered

RobinLinus said the server cost where high especially with my addon which is running in background, so i dont want to just shift that issue to pairdrop thats why i'm also interested in @schlagmichdoch opinion (maybe the heartbeat ping could be modified for the addon so the strain on the server is reduced, what do you think?)

### Additional context

https://github.com/schlagmichdoch/PairDrop

https://pairdrop.net/

**Comment by schlagmichdoch** [comment]
Thanks for putting me in the loop!
There is already an issue (#246) which requests PairDrop support but maybe this new feature is good to discuss coordination whereas the other issue is more about the technicalities. @fm-sys If you don't like to have a duplicate I don't mind moving the coordination part to the PairDrop repo.

I had lots of feature requests for PairDrop so I did not work further on the compatibility with this app, I'm afraid. I will however give it another try soon but this will probably only make the core Snapdrop feature of PairDrop compatible with the app and leave out new features like pairing or perspectively rooms.

Snapdrop for Firefox seems to work though. What do you mean with "just fixing my Addon"?


> RobinLinus said the server cost where high especially with my addon which is running in background

Did RobinLinus say that himself? How did you differentiate traffic via the addon from traffic via any other browser tab?
Why does he / you think the addon is a strain on the server?

I think about moving PairDrop to a VPS without a bandwidth limit which should make this issue not so much relevant ([see this issue and comment](https://github.com/schlagmichdoch/PairDrop/issues/71#issuecomment-1462993400))

> maybe the heartbeat ping could be modified for the addon

I reduced the heartbeat timeout from Snapdrops 30s to 0.5s for stability reasons. Actually, I don't think the heartbeat will be any problem. I just checked and after 60s of keeping PairDrop open about 4.4 KB were transferred via the websocket. If your addon is kept in background for 8h that adds up to only ~2.1 MB. Further, the files used by the addon will probably be cached by Firefox.

**Comment by ueen** [comment]
Yea i updated the snapdrop client js in the add-on but as far as I tried the background service with pairdrop doesn't seem to work because the snapdrop js client is modified in the background service. So to support pairdrip it would just be adding the pairdrop client js to the add-on.

About bandwidth this was said
https://github.com/RobinLinus/snapdrop/pull/580#event-8717224556
But if that's no issue for you that's great! 

I think the main question then is @fm-sys is it better to do a separate app/addon or incorporate compatibility in one project, I m not entirely sure, if one project we would need to differentiate between snapdrop and pairdrop, maybe could take inspiration from your instance verification process?

**Comment by fm-sys** [comment]
@schlagmichdoch 

I have _many_ bad reviews cause snapdrop.net is so unstable. 

According to the Google play console, Snapdrop for Android currently has 500k active installations, probably <1% of it is using the app in peak times. I'm thinking about moving the default URL to a different instance, especially as RobinLinus doesn't seem to have much spare time for maintenance and I never wanted to crash his servers by causing too much load. (I cannot say how much load is actually caused by my app)

Would it be ok for you to use pairdrop as the default instance and could your severs handle the load? However I don't want to cause high additional costs for you...

However, we need to find a way to stay compatible with classic snapdrop instances

**Comment by schlagmichdoch** [comment]
I would like that. Maybe it's a little confusing for your users if they need to use pairdrop.net if the app is called Snapdrop for Android. Maybe you could also simply adapt your "Server down" message. If the current base url is snapdrop.net you could offer a button to change the base url to pairdrop.net. That way the transition is optional and the additional charge onto the server is not all at once.
For new users a separate app called PairDrop for Android would probably be easier to understand but I don't know.
I guess for now PairDrop and Snapdrop are simililar enough to achieve compatibility with PairDrop and Snapdrop instances simultaneously.

Currently I'm not hosting pairdrop.net on a VPS and there is a bandwidth limit that will be reached if the current monthly userbase increases by 400%. As I told @ueen I think about moving it to a VPS/Root server mainly because of a needed TURN server for some connections. I'm not entirely sure about scaling.

## RAM/CPU load
I guess most servers should handle the RAM / CPU load as it is mostly websockets that are handled. Even 100,000 websocket connections simultaneously should be manageable with some GB of RAM: https://socket.io/docs/v4/memory-usage/#memory-usage-per-websocket-server-implementation (see green line).

## Network speed
If 1 websocket connection uses `<10 KB/min` < `<0.167 KB/s` = `<0.001336 Mbit/s`, then `1000 Mbit/s` should be enough for 748,502 simultaneous connections. If that's not enough in the long run the server should be updated to 2500 Mbit/s.

## Bandwidth
If there will be 500,000 users every day with 1 MB of bandwidth per session we would have a monthly bandwidth of 31 * 500,000 * 1 MB = 15,500,000 MB = 15.5 TB
As I also need to add a TURN server to relay some of the traffic itself, an often offered bandwidth limit of 80 TB probably won't be enough.

What do you think? Is this completely irrational or more or less realistic?

I will probably start small and cheap anyways and upgrade as necessary.

**Comment by schlagmichdoch** [comment]
The node.js server should probably also be changed to enable more load: https://blog.jayway.com/2015/04/13/600k-concurrent-websocket-connections-on-aws-using-node-js/

**Comment by fm-sys** [comment]
> What do you think? Is this completely irrational or more or less realistic?

Leaving out the turn server would dramatically reduce load for sure.

You had mentioned 2MB/8h, so I assume that load caused by this app will be significantly less than 1 MB/day (without turn Server). OTOH we have to calculate with some bandwidth for loading the html files etc, so 1 MB/user/day might be realistic.

> Maybe it's a little confusing for your users if they need to use pairdrop.net if the app is called Snapdrop for Android.

I'm thinking of a small instalation wizard offering to select between snapdrop.net, pairdrop and a custom URL. IMHO this should be understandable for users...

**Comment by ueen** [comment]
Maybe just rename to “Snapdrop / Pairdrop“ in the playstore (the "for Android" seems redundant there anyway :)

**Comment by schlagmichdoch** [comment]
Thanks @fm-sys and @ueen ! Snapdrop for Android (master) and PairDrop `v1.5.0` seem to be compatible with each other 🎉

I have added _Snapdrop for Android_ and _Snapdrop for Firefox_ to the Third-Party list in the FAQ: https://github.com/schlagmichdoch/PairDrop/blob/master/docs/faq.md#frequently-asked-questions

**Comment by ueen** [comment]
Awesome, thanks for the fast resolution! 
The add-on background service doesn't work as of now but I will look into it and maybe create a separate issue @ pairdrop repo :)

**Comment by fm-sys** [comment]
Sneak Preview:

I've started developing a new onboarding experience...

https://user-images.githubusercontent.com/64581222/225169691-6609dd21-7bef-4970-bc7b-288827ead954.mp4

**Comment by ueen** [comment]
Great looks really nice :)

**Comment by fm-sys** [comment]
Onboarding experience would be ready from my side. I would appreciate if you could take a closer look (e.g. typos, string improvement ideas, etc.)

https://user-images.githubusercontent.com/64581222/225273896-ab74c2ec-a0cc-43cb-bb57-769c8caf10e4.mp4

**Comment by ueen** [comment]
Hmm don't love the long onboarding, is the permission really necessary, or worked without before? The tutorial page is nice but imho the snapdrop design is/should be intuitive enough, maybe add it to the main views top bar with a ?-icon instead?

**Comment by ueen** [comment]
Another idea: show the url in the top bar (instead of "snapdrop for android" so it will be easier too enter the same address on other devices :)

**Comment by fm-sys** [comment]
> Hmm don't love the long onboarding, is the permission really necessary, it worked without before?

In the video I denied the permission first. That's why it is shown two times. 

We need to ask for storage permission prior to Android 10. This page will be skipped on Android 10 and above. (that's the same way as it was implemented before, only UI is different) 

> The tutorial page is nice but imho the snapdrop design is/should be intuitive enough, 

From time to time users ask "why there is no IOS app" or "would be perfect if Windows is supported as well". Currently it's not recognizable that snapdrop actually is a web-based application. I tried to make this clear inside the last onbording screen, but probably the hint is still to small. 

What about showing an image of a browser URL bar on the last screen? 

> Another idea: show the url in the top bar (instead of "snapdrop for android" so it will be easier too enter the same address on other devices :)

Branding is important as well 😂

But we could show the URL e.g. as actionbar subtitle

**Comment by ueen** [comment]
I wouldn't use "instance" that's to technical for the average user, even server might be to much. Users are familiar with websites and (URL) addresses so that language should be used (while being accurate in the description).

**Comment by schlagmichdoch** [comment]
True. I would go with "server address" then. That's what nextcloud is using.
https://github.com/nextcloud/android

**Comment by fm-sys** [comment]
> having said that, action subtitle might just work too :)

![grafik](https://user-images.githubusercontent.com/64581222/225355223-b5d9cdc6-4781-4317-adae-9fa6f6efaac8.png)

**Comment by fm-sys** [comment]
> consider renaming the playstore application to "Snapdrop / Pairdrop"

Need to think about it. Would require to redo all feature graphics etc. Actually I would like to avoid that effort...

**Comment by ueen** [comment]
> Need to think about it. Would require to redo all feature graphics etc. Actually I would like to avoid that effort...

Yea that would be annoying, but I think the app itself doesn't need to be too much of a brand, the brand (if you could even call it that) is snapdrop or pairdrop, so that's what users will search for and with that should be able to find the app easily, its a strongly practical approach, I know 😇

Thanks for the great work, really looking forward to using the new version 😃

**Comment by fm-sys** [comment]
I'll stay with the last onboarding screen, however, I changed the screenshot to make the browser feature more recognizable... Don't worry, you'll see it only once ;)

![grafik](https://user-images.githubusercontent.com/64581222/225374343-ed506bbb-1493-4955-873b-4495ff20f0f3.png)

**Comment by fm-sys** [comment]
> Currently I'm not hosting pairdrop.net on a VPS and there is a bandwidth limit that will be reached if the current monthly userbase increases by 400%.

@schlagmichdoch I plan to release a new version of Snapdrop for Android soon. I plan to use a scheduled rollout and only target a small amount of users for now. If nothing serious occurs, I'll raise the percentage over time. Please give me the go if you're ready, or tell me whether you want to make some adaptions at the server backend first.

**Comment by LisaJackson24-2** [comment]
> > Currently I'm not hosting pairdrop.net on a VPS and there is a bandwidth limit that will be reached if the current monthly userbase increases by 400%.
> 
> @schlagmichdoch I plan to release a new version of Snapdrop for Android soon. I plan to use a scheduled rollout and only target a small amount of users for now. If nothing serious occurs, I'll raise the percentage over time. Please give me the go if you're ready, or tell me whether you want to make some adaptions at the server backend first.

Please do not combine Snapdrop and PairDrop into one app. If you want it so badly, just split them.

PairDrop is the dumbest piece of crap I've ever seen. The devices either do not connect, or the transfer of a file of several megabytes takes forever.

**Comment by fm-sys** [comment]
> PairDrop is the dumbest piece of crap I've ever seen. 

This might be your personal opinion, please try to stay polite anyway. FYI: You will still be able to select Snapdrop, PairDrop or a custom URL. This selection gets persisted, so you won't see PairDrop again if you don't like to.

**Comment by schlagmichdoch** [comment]
> Please give me the go if you're ready

I’ll need some days to set everything up to be ready if the user base increases and for the turn server but I guess you could start releasing the new version whenever you’re ready as everything works fine.

> snapdrop or pairdrop, so that's what users will search for and with that should be able to find the app easily

Probably tagging would be enough for now. Users also find it in the faq so I guess that’s not a problem for now. 

@LisaJackson24-2 if you have any problem with PairDrop please create an issue on its repository. It is a fork of Snapdrop and the underlying connection code has not changed a lot so speed and connection stability should be the same. The thing that has changed though is that devices are only shown if the connection is successfully established. With Snapdrop many people had the problem that devices were shown but the actual p2p connection failed so that no transfer was possible. If you have suggestions how to improve the connection or the UI you are welcome to create a request.

**Comment by fm-sys** [comment]
> I’ll need some days to set everything up to be ready if the user base increases and for the turn server but I guess you could start releasing the new version whenever you’re ready as everything works fine.

Ok, started roll out to 20% of our users. I'll plan to raise it to 100% starting next week... Then it will take some more weeks until the update is actually installed on all devices.

**Comment by deafmute1** [comment]
@fm-sys Given that pairdrop allows connections between paired devices across different networks, the "Please verify you are connected with a WiFi network" pop up is now redundant in some cases, because pairdrop can still work with a paired device if my phone is on cellular. This pop up should either be disabled for pairdrop, or there should be at least an option to disable it in the settings.

**Comment by fm-sys** [comment]
Let's close here as the new version should be available to all users meanwhile. Everything else should go into dedicated follow up issues IMHO...

@deafmute1 Good catch, I've opened #288 for it

**Comment by fm-sys** [comment]
@schlagmichdoch

Finally I followed some of your suggestions. The main activity title "Snapdrop for Android" is gone meanwhile and the app is called "Snapdrop & Pairdrop" in the playstore. It turned out that more users dare to use Pairdrop if it's explicitly mentioned in the app name.

However I have following problem: How should we name the launcher icon? Any ideas for a Snapdrop / PairDrop independ name?


Cause "Snapdrop" doesn't fit anymore if we actually recommend PairDrop as server address... 

![Screenshot_20230525_170207_com huawei android launcher](https://github.com/fm-sys/snapdrop-android/assets/64581222/522882e6-696a-4457-a930-5be5d0ba60a4)

**Comment by schlagmichdoch** [comment]
> Finally I followed some of your suggestions. The main activity title "Snapdrop for Android" is gone meanwhile and the app is called "Snapdrop & Pairdrop" in the playstore. It turned out that more users dare to use Pairdrop if it's explicitly mentioned in the app name.

I believe the description is a big part too but the changes you mention sound good! 👍 

> How should we name the launcher icon? Any ideas for a Snapdrop / PairDrop independ name?

Not sure. Is there a maximum lenght that is recommended? Otherwise your could go with `Snapdrop & PairDrop`

I have one hacky Idea:
You could just use `Drop` as the launcher name and change the launcher icon depended on the server selection:
![PairDropHacky](https://github.com/fm-sys/snapdrop-android/assets/52242352/9af0a27b-fffc-4f6c-8631-2db8f589efea)

See [here](https://pub.dev/packages/flutter_dynamic_icon) and [here](https://medium.com/flutter-community/programatically-change-ios-app-icon-in-flutter-c9e84bc541a2).

But I guess that would look cheap on some devices with different system fonts.

Nevertheless, just "Drop" would maybe be sufficient too :)

You could also go all in with PairDrop but that would simply reverse the problem. 😉

**Comment by fm-sys** [comment]
> You could just use `Drop` as the launcher name and change the launcher icon depended on the server selection

In the docs they state _Supports only iOS (with version > 10.3)._

So won't work for Android devices unfortunately, and would look weird anyway.

> Is there a maximum lenght that is recommended? Otherwise your could go with Snapdrop & PairDrop

 Not sure but launchers will truncate the name if it's too long.

I'll probably just wait some more time. Let's see how things will evolve.
